### PR TITLE
Introduce `multilinestring_range` structure, simplifies point-linestring distance API

### LINKED ISSUE #705: Simplify header-only point_linestring_distance API
Adopt iterator collection structure to simplify the interface.

### PR DESCRIPTION
**Pull request by isVoid**
## Description

Follow up to #731 , this PR introduces a new range object named `multilinestring_range` and simplifies the point-linestring distance API. Closes #705 .

## Checklist
- [x] I am familiar with the [Contributing Guidelines](https://github.com/rapidsai/cuspatial/blob/HEAD/CONTRIBUTING.md).
- [x] New or existing tests cover these changes.
- [x] The documentation is up to date with these changes.

### CODE REVIEW

File: cpp/include/cuspatial/experimental/detail/ranges/multipoint_range.cuh [hunk 1]
```diff
@@ -101,7 +101,8 @@ template <typename IndexType>
 CUSPATIAL_HOST_DEVICE auto multipoint_range<GeometryIterator, VecIterator>::operator[](
   IndexType idx)
 {
-  return multipoint_begin()[idx];
+  return multipoint_ref<VecIterator>{_points_begin + _geometry_begin[idx],
+                                     _points_begin + _geometry_begin[idx + 1]};
```
Comment (isVoid) [inline cpp/include/cuspatial/experimental/detail/ranges/multipoint_range.cuh:105]: I had to revert this change in #731 because I need to use this method in device code. The previous proposal requires creating a transform iterator and dereference it. Not all code path is available on device and is expensive.

File: cpp/include/cuspatial/experimental/detail/point_linestring_distance.cuh [hunk 2]
```diff
@@ -41,176 +41,80 @@ namespace cuspatial {
...
-    vec_2d<T> const b      = linestring_points_first[idx + 1];
+    auto [a, b]            = multilinestrings.segment(idx);
     T min_distance_squared = std::numeric_limits<T>::max();
 
-    for (auto point_idx = point_geometry_offset_first[geometry_idx];
-         point_idx < point_geometry_offset_first[geometry_idx + 1];
-         point_idx++) {
-      vec_2d<T> const c = points_first[point_idx];
-
+    for (vec_2d<T> const& c : multipoints[geometry_idx]) {
```
Comment (harrism) [inline cpp/include/cuspatial/experimental/detail/point_linestring_distance.cuh:72]: This is so cool!

File: cpp/include/cuspatial/experimental/detail/point_linestring_distance.cuh [hunk 3]
```diff
@@ -41,176 +41,80 @@ namespace cuspatial {
...
       // TODO: reduce redundant computation only related to `a`, `b` in this helper.
       auto const distance_squared = point_to_segment_distance_squared(c, a, b);
-      min_distance_squared        = std::min(distance_squared, min_distance_squared);
+      min_distance_squared        = min(distance_squared, min_distance_squared);
     }
-
     atomicMin(&thrust::raw_reference_cast(*(distances + geometry_idx)),
```
Comment (harrism) [inline cpp/include/cuspatial/experimental/detail/point_linestring_distance.cuh:77]: Can this be simplified? 

```suggestion
    atomicMin(&distances[geometry_idx],
```

File: cpp/include/cuspatial/experimental/detail/point_linestring_distance.cuh [hunk 4]
```diff
@@ -41,176 +41,80 @@ namespace cuspatial {
...
+OutputIt pairwise_point_linestring_distance(MultiPointRange multipoints,
+                                            MultiLinestringRange multilinestrings,
                                             OutputIt distances_first,
                                             rmm::cuda_stream_view stream)
 {
-  using T = iterator_vec_base_type<Cart2dItA>;
+  using T = typename MultiPointRange::element_t;
 
-  static_assert(is_same_floating_point<T, iterator_vec_base_type<Cart2dItB>>(),
+  static_assert(is_same_floating_point<T, typename MultiLinestringRange::element_t>(),
```
Comment (harrism) [inline cpp/include/cuspatial/experimental/detail/point_linestring_distance.cuh:91]: Isn't this assertion covered by the `is_same` assertion below? If the vec_2ds are the same type, then the T must be the same as well.

File: cpp/include/cuspatial/experimental/detail/point_linestring_distance.cuh [hunk 5]
```diff
@@ -41,176 +41,80 @@ namespace cuspatial {
...
-      points_first,
-      points_last,
-      linestring_geometry_offset_first,
-      linestring_geometry_offset_first + num_pairs + 1,
-      linestring_part_offsets_first,
-      linestring_part_offsets_last,
-      linestring_points_first,
-      linestring_points_last,
-      distances_first);
+      multipoints, multilinestrings, distances_first);
```
Comment (harrism) [inline cpp/include/cuspatial/experimental/detail/point_linestring_distance.cuh:113]: So nice.

File: cpp/include/cuspatial/experimental/ranges/multilinestring_range.cuh [hunk 6]
```diff
@@ -0,0 +1,178 @@
...
+#include <cuspatial/cuda_utils.hpp>
+#include <cuspatial/traits.hpp>
+#include <cuspatial/vec_2d.hpp>
+
+namespace cuspatial {
+
+/**
+ * @brief Non-owning object of a multilinestring array
+ * @ingroup ranges
+ *
```
Comment (harrism) [inline cpp/include/cuspatial/experimental/ranges/multilinestring_range.cuh:30]: Suggest a little more explanation. E.g.

Provides a range-based interface to contiguous storage of multilinestring data, to make it easier to access and 
iterate over linestrings and points within linestrings.

File: cpp/include/cuspatial/experimental/ranges/multilinestring_range.cuh [hunk 7]
```diff
@@ -0,0 +1,178 @@
...
+#include <thrust/pair.h>
+
+#include <cuspatial/cuda_utils.hpp>
+#include <cuspatial/traits.hpp>
+#include <cuspatial/vec_2d.hpp>
+
+namespace cuspatial {
+
+/**
+ * @brief Non-owning object of a multilinestring array
```
Comment (harrism) [inline cpp/include/cuspatial/experimental/ranges/multilinestring_range.cuh:28]:
```suggestion
 * @brief Non-owning range-based interface to multilinestring data.
```

File: cpp/include/cuspatial/experimental/ranges/multilinestring_range.cuh [hunk 8]
```diff
@@ -0,0 +1,178 @@
...
+
+  /**
+   * @brief Given the index of a part (linestring), return the geometry (multilinestring) index
+   * where the linestring locates.
+   */
+  template <typename IndexType>
+  CUSPATIAL_HOST_DEVICE auto geometry_idx_from_part_idx(IndexType part_idx);
+
+  /**
+   * @brief Given the index of the point, return the geometry (multilinestring) index where the
```
Comment (harrism) [inline cpp/include/cuspatial/experimental/ranges/multilinestring_range.cuh:97]:
```suggestion
   * @brief Given the index of a point, return the geometry (multilinestring) index where the
```

File: cpp/include/cuspatial/experimental/ranges/multilinestring_range.cuh [hunk 9]
```diff
@@ -0,0 +1,178 @@
...
+   * point locates.
+   */
+  template <typename IndexType>
+  CUSPATIAL_HOST_DEVICE auto geometry_idx_from_point_idx(IndexType point_idx);
+
+  /**
+   * @brief Given an index of a segment, returns true if the index is valid.
+   * The index of a segment is the same as the index to the starting point of the segment.
+   * Thus, the index to the last point of a linestring is an invalid index for segment.
```
Comment (harrism) [inline cpp/include/cuspatial/experimental/ranges/multilinestring_range.cuh:106]:
```suggestion
   * Thus, the index to the last point of a linestring is an invalid segment index.
```

File: cpp/include/cuspatial/experimental/detail/ranges/multilinestring_range.cuh [hunk 10]
```diff
@@ -0,0 +1,132 @@
...
+  return geometry_idx_from_part_idx(part_idx_from_point_idx(point_idx));
+}
+
+template <typename GeometryIterator, typename PartIterator, typename VecIterator>
+template <typename IndexType1, typename IndexType2>
+CUSPATIAL_HOST_DEVICE bool
+multilinestring_range<GeometryIterator, PartIterator, VecIterator>::is_valid_segment_id(
+  IndexType1 segment_idx, IndexType2 part_idx)
+{
+  return segment_idx != num_points() && segment_idx != (part_begin[part_idx + 1] - 1);
```
Comment (harrism) [inline cpp/include/cuspatial/experimental/detail/ranges/multilinestring_range.cuh:119]: Seems like segment_idx >= num_points() is also not valid. And what about negative indices?

```suggestion
  return segment_idx >= 0 && segment_idx < num_points() && segment_idx < (part_begin[part_idx + 1] - 1);
```

File: cpp/include/cuspatial/experimental/point_linestring_distance.cuh [hunk 11]
```diff
@@ -21,65 +21,24 @@
...
- * geometry of each pair, the end range is implied by linestring_geometry_offset_first +
- * std::distance(`point_geometry_offset_first`, `point_geometry_offset_last`)
- * @param linestring_offsets_first beginning of the range of offsets to the starting point
- * of each linestring
- * @param linestring_offsets_last end of the range of offsets to the starting point
- * of each linestring
- * @param linestring_points_first beginning of the range of linestring points
- * @param linestring_points_last end of the range of linestring points
- * @param distances_first beginning of the output range of distances
+ * @param multipoints Range object of a multipoint array
```
Comment (harrism) [inline cpp/include/cuspatial/experimental/point_linestring_distance.cuh:30]:
```suggestion
 * @param multipoints The range of multipoints, one per computed distance pair..
```

File: cpp/include/cuspatial/experimental/point_linestring_distance.cuh [hunk 12]
```diff
@@ -21,65 +21,24 @@
...
+ * @param multilinestrings Range object of a multilinestring array
```
Comment (harrism) [inline cpp/include/cuspatial/experimental/point_linestring_distance.cuh:31]:
```suggestion
 * @param multilinestrings The range of multilinestrings, one per computed distance pair.
```

File: cpp/include/cuspatial/experimental/detail/point_linestring_distance.cuh (hunk 4 above)
Comment (isVoid) [reply cpp/include/cuspatial/experimental/detail/point_linestring_distance.cuh:91]: The one below doesn't test floating points type.

File: cpp/include/cuspatial/experimental/detail/ranges/multilinestring_range.cuh (hunk 10 above)
Comment (isVoid) [reply cpp/include/cuspatial/experimental/detail/ranges/multilinestring_range.cuh:119]: Ok that's right. I wasn't thinking the use of this API outside of the context where it was originally used.

Comment (isVoid) [reply cpp/include/cuspatial/experimental/detail/ranges/multilinestring_range.cuh:119]: I think `segment_idx < (part_begin[part_idx + 1] - 1)` is stronger than `segment_idx < num_points()`. Because in geoarrow format the last element of the offset array is the number of element in the next level.

File: cpp/include/cuspatial/experimental/detail/point_linestring_distance.cuh (hunk 3 above)
Comment (isVoid) [reply cpp/include/cuspatial/experimental/detail/point_linestring_distance.cuh]: Yes*
But not without this support: https://github.com/isVoid/cuspatial/blob/17b99149ccfbf85bd2abfeeaf8329f2de70cfe02/cpp/include/cuspatial/detail/utility/device_atomics.cuh#L184-L186

Comment (isVoid) [reply cpp/include/cuspatial/experimental/detail/point_linestring_distance.cuh]: This is because `distances[geometry_idx]` can be many things. If it's a `float` then it's all good. Other times it can be `thrust::device_reference`, which would require some extra work to cast that into a raw pointer. The easiest of all approach is just to override atomic to accept `thrust::device_ptr`. Let's hope `atomic_ref` plays nicely with device vectors in the future.

File: cpp/include/cuspatial/experimental/ranges/multilinestring_range.cuh [hunk 13]
```diff
@@ -0,0 +1,181 @@
...
+#include <thrust/pair.h>
+
+#include <cuspatial/cuda_utils.hpp>
+#include <cuspatial/traits.hpp>
+#include <cuspatial/vec_2d.hpp>
+
+namespace cuspatial {
+
+/**
+ * @brief Non-owning range-based interface to a multilinestring data
```
Comment (harrism) [inline cpp/include/cuspatial/experimental/ranges/multilinestring_range.cuh:28]:
```suggestion
 * @brief Non-owning range-based interface to multilinestring data
```

Comment (harrism) [reply cpp/include/cuspatial/experimental/ranges/multilinestring_range.cuh:28]: Sorry for the one-character suggestion. (But it matches the multipoint comment this way)

**Review by harrism — APPROVED** [review]
Looks good!

**Review by thomcom — CHANGES_REQUESTED** [review]
Cool!

File: cpp/include/cuspatial/experimental/ranges/multilinestring_range.cuh [hunk 14]
```diff
@@ -0,0 +1,181 @@
...
+                                IndexType3 num_points,
+                                VecIterator point_begin)
+{
+  return multilinestring_range{geometry_begin,
+                               geometry_begin + num_multilinestrings + 1,
+                               part_begin,
+                               part_begin + num_linestrings + 1,
+                               point_begin,
+                               point_begin + num_points};
+}
```
[the remark this answers is not in the document]
Comment (isVoid) [reply cpp/include/cuspatial/experimental/ranges/multilinestring_range.cuh:177]: Very useful :) I would call it `make_multilinestring_range_from_span`

File: cpp/tests/experimental/spatial/point_linestring_distance_test.cu [hunk 15]
```diff
@@ -82,31 +84,32 @@ TYPED_TEST(PairwisePointLinestringDistanceTest, OnePairFromVectorSingleComponent
...
 
+  auto multipoints = make_multipoint_range(
+    d_point_geometries.size() - 1, d_point_geometries.begin(), d_points.size(), d_points.begin());
+
+  auto multilinestrings = make_multilinestring_range(d_linestring_geometries.size() - 1,
+                                                     d_linestring_geometries.begin(),
+                                                     d_linestring_parts.size() - 1,
+                                                     d_linestring_parts.begin(),
+                                                     d_linestring_points.size(),
+                                                     d_linestring_points.begin());
```
[the remark this answers is not in the document]
Comment (isVoid) [reply cpp/tests/experimental/spatial/point_linestring_distance_test.cu:108]: Will do!

File: cpp/include/cuspatial/experimental/ranges/multilinestring_range.cuh [hunk 16]
```diff
@@ -0,0 +1,181 @@
...
+ * @param num_points Number of points in the underlying points array.
+ * @param point_begin Iterator to the start of the point array.
+ * @return A `multilinestring_range` object
+ *
+ * [LinkLRAI]: https://en.cppreference.com/w/cpp/named_req/RandomAccessIterator
+ * "LegacyRandomAccessIterator"
+ */
+template <typename IndexType1,
+          typename IndexType2,
+          typename IndexType3,
```
[the remark this answers is not in the document]
Comment (isVoid) [reply cpp/include/cuspatial/experimental/ranges/multilinestring_range.cuh:160]: Sure.

File: cpp/include/cuspatial/experimental/ranges/multilinestring_range.cuh (hunk 14 above)
[the remark this answers is not in the document]
Comment (isVoid) [reply cpp/include/cuspatial/experimental/ranges/multilinestring_range.cuh:177]: You also don't need to use size() here, you can just directly call the constructor using begin and end from the buffers.

File: cpp/include/cuspatial/experimental/detail/point_linestring_distance.cuh [hunk 17]
```diff
@@ -41,176 +41,79 @@ namespace cuspatial {
...
+OutputIt pairwise_point_linestring_distance(MultiPointRange multipoints,
+                                            MultiLinestringRange multilinestrings,
                                             OutputIt distances_first,
                                             rmm::cuda_stream_view stream)
 {
-  using T = iterator_vec_base_type<Cart2dItA>;
+  using T = typename MultiPointRange::element_t;
 
-  static_assert(is_same_floating_point<T, iterator_vec_base_type<Cart2dItB>>(),
+  static_assert(is_same_floating_point<T, typename MultiLinestringRange::element_t>(),
```
[the remark this answers is not in the document]
Comment (isVoid) [reply cpp/include/cuspatial/experimental/detail/point_linestring_distance.cuh:90]: `T` is `MultiPointRange::element_t`. This constraints the following two things:
1. `MultiLinestringRange::element_t` and T is the same
2. They are both floating point types.

File: cpp/include/cuspatial/experimental/ranges/multilinestring_range.cuh (hunk 14 above)
[the remark this answers is not in the document]
Comment (isVoid) [reply cpp/include/cuspatial/experimental/ranges/multilinestring_range.cuh:177]: I added the overload, though I didn't bother to replace all usage of the other factory since the rest are working quite well, except for one just for testing purposes.

**Review by thomcom — APPROVED** [review]

**Comment by thomcom** [comment]
Approvin

**Comment by isVoid** [comment]
@gpucibot merge

**Comment by isVoid** [comment]
@gpucibot merge
